prow/cmd/webhook-server/main.go

```diff
@@ -48,118 +47,158 @@ type ClientInterface interface {
-		logrus.WithError(err).Fatal("Unable to create temp directory")
-	}
-	defer os.RemoveAll(tempDir)
-	certFile := filepath.Join(tempDir, "certFile.pem")
-	if err := ioutil.WriteFile(certFile, []byte(cert), 0666); err != nil {
-		logrus.WithError(err).Fatal("Could not write contents of cert file")
-	}
-	privKeyFile := filepath.Join(tempDir, "privKey.pem")
-	if err := ioutil.WriteFile(privKeyFile, []byte(privKey), 0666); err != nil {
-		logrus.WithError(err).Fatal("Could not write contents of privKey file")
+		defer os.RemoveAll(tempDir)
+		certFile = filepath.Join(tempDir, "certFile.pem")
+		if err := ioutil.WriteFile(certFile, []byte(cert), 0666); err != nil {
+			logrus.WithError(err).Fatal("Could not write contents of cert file")
+		}
+		privKeyFile = filepath.Join(tempDir, "privKey.pem")
+		if err := ioutil.WriteFile(privKeyFile, []byte(privKey), 0666); err != nil {
+			logrus.WithError(err).Fatal("Could not write contents of privKey file")
+		}
+	} else {
+		fs := FileSystemClient{}
+		absPath, err := filepath.Abs(fsSecretPath)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to generate absolute path")
+		}
+		certFile = filepath.Join(absPath, "certFile.pem")
+		privKeyFile = filepath.Join(absPath, "privKeyFile.pem")
+		caBundleFile := filepath.Join(absPath, "caBundle.pem")
+		secretFiles := []string{certFile, privKeyFile, caBundleFile}
+		for _, secretFile := range secretFiles {
+			_, err := ioutil.ReadFile(secretFile)
+			if err != nil {
+				err := fs.CreateSecret(ctx, secretID)
+				if err != nil {
+					logrus.WithError(err).Fatal("Could not create ca-secret directory")
+				}
+				fs.AddSecretVersion(absPath, ctx, secretID, o.expiryInYears, o.dnsNames)
+				
+				caPem, err := ioutil.ReadFile(caBundleFile)
+				if err != nil {
+					logrus.WithError(err).Fatal("Unable to read contents of caBundle file")
+				}
+				err = createValidatingWebhookConfig(ctx, string(caPem), cl)
+				if err != nil && strings.Contains(err.Error(), configAlreadyExistsError) {
+					err = patchValidatingWebhookConfig(ctx, string(caPem), cl, true)
+					if err != nil {
+						logrus.WithError(err).Fatal("Unable to patch ValidationWebhookConfig")
+					}
+				} else if err != nil {
+					logrus.WithError(err).Fatal("Unable to create ValidationWebhookConfig")
+				}
+				break
+			} 
+		}
+		cert, _ := ioutil.ReadFile(certFile)
+		if err = isCertValid(string(cert)); err != nil {
+			logrus.WithError(err).Info("Certificate is not valid, will replace.")
+			err := fs.AddSecretVersion(absPath, ctx, secretID, o.expiryInYears, o.dnsNames)
+			if err != nil {
+				logrus.WithError(err).Fatal("Could not add secret version")
+			}
+			caPem, err := ioutil.ReadFile(caBundleFile)
+			if err != nil {
+				logrus.WithError(err).Fatal("Unable to read contents of caBundle file")
+			}
+			err = patchValidatingWebhookConfig(ctx, string(caPem), cl, true)
+			if err != nil {
+				logrus.WithError(err).Fatal("Unable to generate ValidationWebhookConfig")
+			}
+		} 
 	}
```

Comment:
not entirely sure that it's necessary to duplicate this logic. What's blocking us from unifying to be a single function and using ClientInterface when there is a difference between GCP and local FS? In my opinion there should be no `if else`, and the logic should be:
```
var client ClientInterface
if o.Project {
  client = Client for GCP
}
if o.localFS {
  client = Client for local FS
}

// Use client whenever need to access the secret
```